### ISSUE TITLE
GovCloud Signalfx integrations

### ISSUE THREAD

**Issue by rss04** [splunk-terraform/terraform-provider-signalfx#394]
When using this code with govcvloud, I am receiving the following error :

```
Error: Invalid address to set: []string{"external_id"}

```

Following code was used:
```
resource "signalfx_aws_integration" "aws_idemia_integration_st" {
  count          = var.use_security_token ? 1 : 0
  named_token    = signalfx_org_token.aws_integration.name
  integration_id = signalfx_aws_token_integration.aws_token.id

  #if auth_method = \"SecurityToken\"
  token = aws_iam_access_key.splunk_obs_iam_user_access_key[0].id
  key   = aws_iam_access_key.splunk_obs_iam_user_access_key[0].secret


  enable_aws_usage          = var.enable_aws_usage
  enable_check_large_volume = var.enable_check_large_volume
  enable_logs_sync          = var.enable_logs_sync
  enabled                   = var.enable_integration #defaults is true

  #custom_cloudwatch_namespaces = var.custom_cloudwatch_namespaces
  dynamic "custom_namespace_sync_rule" {
    for_each = var.custom_namespace_sync_rules
    content {
      default_action = try(custom_namespace_sync_rules.value["default_action"], "Exclude")
      filter_action  = try(custom_namespace_sync_rules.value["filter_action"], "Include")
      filter_source  = try(custom_namespace_sync_rules.value["filter_source"], "filter('code', '200')")
      namespace      = try(var.custom_namespace_sync_rules.value["namespace"], "Idemia-custom-namespace")
    }
  }
  sync_custom_namespaces_only = var.sync_custom_namespaces_only #defaults to false
  import_cloud_watch          = var.import_cloud_watch          #default is true
  dynamic "metric_stats_to_sync" {
    for_each = var.metric_stats_to_sync
    content {
      metric    = try(metric_stats_to_sync.value["metric"], "NetworkPacketsIn")
      namespace = try(metric_stats_to_sync.value["namespace"], "Idemia-custom-namespace")
      stats     = try(metric_stats_to_sync.value["stats"], ["upper"])
    }
  }
  #name        = var.integration_name

  dynamic "namespace_sync_rule" {
    for_each = var.monitor_aws_services ? [] : var.namespace_sync_rules
    content {
      default_action = try(custom_namespace_sync_rules.value["default_action"], "Exclude")
      filter_action  = try(custom_namespace_sync_rules.value["filter_action"], "Include")
      filter_source  = try(custom_namespace_sync_rules.value["filter_source"], "filter('code', '200')")
      namespace      = try(var.custom_namespace_sync_rules.value["namespace"], "AWS/EC2")
    }
  }
  poll_rate                  = var.poll_rate #default is 300
  regions                    = var.regions
  role_arn                   = var.use_security_token ? null : aws_iam_role.aws_splunk_role[0].arn
  use_get_metric_data_method = var.use_get_metric_data_method
  use_metric_streams_sync    = var.use_metric_streams_sync
  services = var.monitor_aws_services

}
```
Can you please advise ?

Thanks

**Comment by dgcom** [comment]
There is a bug in provider which shows its ugly head (`Error: Invalid address to set: []string{"external_id"}`) if `signalfx_aws_token_integration` is created in a separate run from `signalfx_aws_integration`.
Workaround until fixed - make sure that both resources are created in single TF plan/apply run.

Details:
`IntegrationAWSRead` receives `"external_id"` field from API call in `GetAWSCloudWatchIntegration` and attempts to set it in the context which does not have this field defined for token authentication.
This happens in plan stage if `signalfx_aws_token_integration` exists but `signalfx_aws_integration` doesn't.

**Comment by akonarska** [comment]
@rss04 I hope your issue is solved. @dgcom is correct in terms of how to solve it - both resources have to be created in the same run and it is not a bug - this is how our API currently works.

So basically you need to define `signalfx_aws_token_integration` or `signalfx_aws_external_integration` along with `signalfx_aws_integration` - see example here https://registry.terraform.io/providers/splunk-terraform/signalfx/latest/docs/resources/aws_integration.

I will be closing this for now - feel free to reopen in case any more help is needed.

**Comment by dgcom** [comment]
@akonarska The design of this provider is the issue if

> both resources have to be created in the same run

This goes against Terraform principles.

Moreover, there is an easy fix (or, more like a workaround) for this which can be implemented in the code - just add the missing field.

The bigger problem is the API behavior...

Speaking of provider design, here are the guidelines: [Provider Design Principles - Guides | Terraform | HashiCorp Developer](https://developer.hashicorp.com/terraform/plugin/best-practices/hashicorp-provider-design-principles)

> A Terraform resource should be a declarative representation of single component, usually with create, read, delete, and optionally update methods. 

This means that cloud integration should be a single resource and not two dependent resources as it is currently.

**Comment by akonarska** [comment]
@dgcom design of this provider is reflecting the underlying API design and unfortunately that API require 2 calls to create a fully operational cloud integration (see [docs](https://docs.splunk.com/Observability/gdi/get-data-in/connect/aws/aws-apiconfig.html)). That seems to be the best solution to deal with it. 
I agree with you - this is not great and we are very well aware of that. We have it on our roadmap but it will take some time to get it out.

**Comment by akonarska** [comment]
@dgcom It look like I did not understand your concerns fully and closed this issue too eagerly. Sorry for that - it was a bug indeed. We have a fix for it ready (https://github.com/splunk-terraform/terraform-provider-signalfx/pull/445) and it will be released on Monday with `8.0.0` version of the provider. Please let us know in case there is still something not working here. If we don't hear from you in two weeks I will assume that everything works and close this issue.

**Comment by akonarska** [comment]
Version 8.0.0 is now available and it contains fix for this bug. @dgcom @rss04 please check if that works now as expected.

**Comment by akonarska** [comment]
As confirmed by @dgcom - this was fixed in 8.0.0.
